magma/clock.py

```diff
@@ -134,29 +134,51 @@ def first(iterable):
     return next((item for item in iterable if item is not None), None)
 
 
-def get_first_clock(port, clocktype):
+def get_clocks(port, clocktype):
+    if port.is_input():
+        # We are looking for default drivers (outputs)
+        return []
     if isinstance(port, clocktype):
-        return port
+        return [port]
     if isinstance(port, Tuple):
-        clks = (get_first_clock(elem, clocktype) for elem in port)
-        return first(clks)
+        clks = []
+        for elem in port:
+            clks += get_clocks(elem, clocktype)
+        return clks
     if isinstance(port, Array):
-        return get_first_clock(port[0], clocktype)
-    return None
+        clks = get_clocks(port[0], clocktype)
+        if not clks:
+            # early exit to avoid traversing children when unnecessary
+            return clks
+        for elem in port[1:]:
+            clks += get_clocks(elem, clocktype)
+        return clks
+    return []
+
+
+def get_clocks_from_defn(defn, clocktype):
+    clocks = []
+    for port in defn.interface.ports.values():
+        clocks += get_clocks(port, clocktype)
+    for inst in defn.instances:
+        # circular import because Wire depends on clock.py
+        # We could move default clock wiring logic to a different file?
+        from magma.primitives.wire import Wire
+        if isinstance(type(inst), Wire):
+            # Skip clock wires, because they are both inputs and outputs, so we
+            # don't treat them as default drivers
+            continue
+        for port in inst.interface.ports.values():
+            clocks += get_clocks(port, clocktype)
+    return clocks
 
 
 def wireclocktype(defn, inst, clocktype):
     # Check common case: top level clock port
-    clks = (port if isinstance(port, clocktype) else None
-            for port in defn.interface.ports.values())
-    defnclk = first(clks)
-    if defnclk is None:
-        # Check recursive types
-        clks = (get_first_clock(port, clocktype)
-                for port in defn.interface.ports.values())
-        defnclk = first(clks)
-    if defnclk is None:
+    clks = get_clocks_from_defn(defn, clocktype)
+    if len(clks) > 1 or len(clks) == 0:
```

Comment:
if we use a generator above, then we should define a function `only`:

```python
class IterableOnlyException(Exception):
    pass

def only(lst: Iterable):
    it = iter(lst)
    try:
        value = next(it)
    except StopIteration:
        raise IterableOnlyException("Expected one element, got []") from None
    try:
        new_value = next(it)
    except StopIteration:
        return value
    else:
        msg = f"Expected one element got {[value, new_value] + list(it)}"
        raise IterableOnlyException(msg)
```

(feel free to rename `IterableOnlyException` -- it's pretty a terrible name) 

and then we can use it as follows:

```python
clks = get_clocks_from_defn(defn, clocktype)
try:
    defnclk = only(clks)
except IterableOnlyException:
    return
```

Reply:
btw, if possible lets put this `only` function in some common place, perhaps `magma.common`